server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,485 @@
+         */
+        @Nullable // when finishing the transmission of the whole response
+        private ChunkedRestResponseBodyPart bodyPart;
+
+        private boolean isPartComplete;
+        private boolean isLastPart;
+
+        /**
+         * A listener which is created when there are no more available chunks, so transmission is paused, subscribed to in
+         * {@link #getNextPart}, and then completed with the next body part (sequence of zipped chunks, i.e. a new (unique) active
+         * {@link AvailableChunksZipResponseBodyPart}).
+         */
+        private SubscribableListener<ChunkedRestResponseBodyPart> getNextPartListener;
+
+        /**
+         * A collection of {@code Releasable} instances to be released when the next chunk has been fully transmitted. It's a collection
+         * because a chunk may complete several entries, each of which has its own resources to release.
+         */
+        private ArrayList<Releasable> nextReleasables = new ArrayList<>(); // preserved between chunks because will often be unused
+
+        AvailableChunksZipResponseBodyPart(ZipEntry zipEntry, ChunkedRestResponseBodyPart bodyPart) {
+            this.zipEntry = zipEntry;
+            this.bodyPart = bodyPart;
+        }
+
+        @Override
+        public boolean isPartComplete() {
+            return isPartComplete;
+        }
+
+        @Override
+        public boolean isLastPart() {
+            return isLastPart;
+        }
+
+        @Override
+        public void getNextPart(ActionListener<ChunkedRestResponseBodyPart> listener) {
+            assert getNextPartListener != null;
+            getNextPartListener.addListener(listener);
+        }
+
+        /**
+         * Transfer {@link #currentEntryReleasable} into the supplied collection (i.e. add it to {@code releasables} and then clear
+         * {@link #currentEntryReleasable}). Called when the last chunk of the last part of the current entry is serialized, so that we can
+         * start serializing chunks of the next entry straight away whilst delaying the release of the current entry's resources until the
+         * transmission of the chunk that is currently under construction.
+         */
+        private void transferCurrentEntryReleasable(ArrayList<Releasable> releasables) {
+            assert queueRefs.hasReferences();
+
+            if (currentEntryReleasable == null) {
+                return;
+            }
+
+            if (releasables == nextReleasables) {
+                nextReleasables = new ArrayList<>();
+            }
+
+            releasables.add(currentEntryReleasable);
+            currentEntryReleasable = null;
+        }
+
+        @Override
+        public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) throws IOException {
+            assert Transports.isTransportThread(Thread.currentThread());
+
+            final ArrayList<Releasable> releasables = nextReleasables;
```

Comment:
> and if I understand correctly

yes, you do :)

> some kind of lazily created list

I'm pretty sure that would end up being more complex. I added a little more commentary and renamed the field in 88f2ff3f97b5cbf83aa9b7be8d1b91150aa7639c, does that help?

Reply:
EncodeChunk does a few things (while processing the queue)
- populate the bytes stream for the chunk
- potentially accumulate some releasables that need to be released after the chunk is written to the wire
- bundle those up in a releasable bytes reference to be returned

Currently that process mutates global-ish state (e.g. `targetStream`, `nextReleasablesCache`). I think that's the bit I find hard to follow, because until you stare at it and understand what's going on you don't realised that those things are (I think) really local to a single `encodeChunk` iteration, but in a wider scope to accommodate some optimisations. I feel like it'd be a bit clearer if that per-iteration state was bundled into something with an API and it encapsulated all the optimisation. (e.g. call `reset` at the start of the iteration and `createBytesReference` at the end?)

Reply:
I'm hesitant to introduce extra allocations and layers of indirection on hot network write paths if we can help it. The `targetStream` thing is a fairly well-established pattern for this, see `org.elasticsearch.rest.ChunkedRestResponseBodyPart#fromXContent` and `org.elasticsearch.rest.ChunkedRestResponseBodyPart#fromTextChunks`.